File: java/ultimate/MessageHandlerCachingProxy.java
package ultimate;

import java.util.HashMap;
import java.util.Map;

public class MessageHandlerCachingProxy implements MessageHandlerProxy {

    private ThirdPartyMessageHandlerApi thirdPartyMessageHandlerApi;

    private Map<String, String> cache;
    public MessageHandlerCachingProxy(ThirdPartyMessageHandlerApi thirdPartyMessageHandlerApi) {
        this.thirdPartyMessageHandlerApi = thirdPartyMessageHandlerApi;
        this.cache = new HashMap<>();
    }

    public String processMessage(String message) {
        if (!cache.containsKey(message)) {
            cache.put(message, thirdPartyMessageHandlerApi.processMessage(message));
            return cache.get(message);
        }
        Logger.getInstance().logInfo("Getting result from the proxy cache...");
        return cache.get(message);
    }
}

File: typescript/patterns/ultimate-design-pattern-challenge/message-handler-caching-proxy.ts
class MessageHandlerCachingProxy implements MessageHandlerProxy {
  private thirdPartyMessageHandlerApi: ThirdPartyMessageHandlerApi;
  private cache: Map<string, string>;

  constructor(thirdPartyMessageHandlingService: ThirdPartyMessageHandlerApi) {
    this.thirdPartyMessageHandlerApi = thirdPartyMessageHandlingService;
    this.cache = new Map();
  }

  processMessage(messageToProcess: string): string {
    if (!this.cache.has(messageToProcess)) {
      Logger.getInstance().logInfo("Saving message to the proxy cache...");

      const processedMessage =
        this.thirdPartyMessageHandlerApi.processMessage(messageToProcess);
      this.cache.set(messageToProcess, processedMessage);

      return this.cache.get(messageToProcess)!;
    }

    Logger.getInstance().logInfo("Getting result from the proxy cache...");
    return this.cache.get(messageToProcess)!;
  }
}
